sky/serve/spot_policy.py

```diff
@@ -0,0 +1,132 @@
+
+    def move_zone_to_preempted(self, zone: str) -> None:
+        assert zone in self.zone2type
+        self.zone2type[zone] = SpotZoneType.PREEMPTED
+
+    def handle_active(self, zone: str) -> None:
+        self.move_zone_to_active(zone)
+
+    def handle_preemption(self, zone: str) -> None:
+        self.move_zone_to_preempted(zone)
+
+    def clear_preempted_zones(self) -> None:
+        for zone in self.zones:
+            self.move_zone_to_active(zone)
+
+    def active_zones(self) -> List[str]:
+        return [
+            zone for zone, zone_type in self.zone2type.items()
+            if zone_type == SpotZoneType.ACTIVE
+        ]
+
+    def preempted_zones(self) -> List[str]:
+        return [
+            zone for zone, zone_type in self.zone2type.items()
+            if zone_type == SpotZoneType.PREEMPTED
+        ]
+
+    def __repr__(self) -> str:
+        return f'{self.NAME}SpotPlacer()'
+
+    @classmethod
+    def from_spec(cls, spec: 'service_spec.SkyServiceSpec') -> 'SpotPlacer':
+        assert (spec.spot_placer is not None and
+                spec.spot_placer in cls.REGISTRY)
+        return cls.REGISTRY[spec.spot_placer](spec)
+
+
+class DynamicFailoverSpotPlacer(SpotPlacer):
+    """Dynamic failover to an active zone when preempted."""
+    NAME: Optional[str] = 'DynamicFailover'
+
+    def _filter_unvisited_active_zones(
+            self, existing_replicas: List['replica_managers.ReplicaInfo']
+    ) -> List[str]:
+        existing_zones = set()
+        info = None
+        for info in existing_replicas:
+            if not info.is_spot:
+                # filter on demand fallbacks
+                continue
+            if info.zone is not None:
+                existing_zones.add(info.zone)
+            else:
+                logger.error(f'Zone is None for replica {info}')
+
+        unvisited_active_zones = [
+            zone for zone in self.active_zones() if zone not in existing_zones
+        ]
+
+        logger.info(f'existing_zones: {existing_zones}')
+        logger.info(f'unvisited_active_zones: {unvisited_active_zones}')
+        return unvisited_active_zones
+
+    def select(self, existing_replicas: List['replica_managers.ReplicaInfo'],
+               num_zones: int) -> List[str]:
+        # Prevent the case with only one active zones.
+        selected_zones = []
+        if (len(self.active_zones()) <= 1 and len(self.preempted_zones()) > 0):
+            self.clear_preempted_zones()
+        unvisited_active_zones = self._filter_unvisited_active_zones(
+            existing_replicas)
+        if unvisited_active_zones:
+            selected_zones.extend(
+                random.sample(unvisited_active_zones,
+                              min(num_zones, len(unvisited_active_zones))))
+            num_zones -= len(selected_zones)
+        if num_zones > 0:
+            selected_zones.extend(
+                random.choices(self.active_zones(), k=num_zones))
```

Comment:
Should we keep track of how many existing replica we have in each active zone and try to spread in those active zone evenly?

Reply:
I think we can, but not sure if this will help too much - randomly choosing from active zones should be quite similar to (static?) evenly spreading among active zones

Reply:
yes I mean randomly picking from active zones will not = evenly spreading. I am not sure if the total preemption experienced in the long run of these two approaches will be significantly different.

Reply:
Pushed a version where the policy tries to maintain an even spread - we should discuss this a bit more.

Reply:
Sure. Lets bring this to our group meeting